pkg/cachemanager/cachemanager.go

```diff
@@ -0,0 +1,385 @@
+		return nil
+	}
+
+	syncKey := syncutil.GetKeyForSyncMetrics(instance.GetNamespace(), instance.GetName())
+	if c.watchesGVK(gvk) {
+		_, err = c.cfClient.AddData(ctx, instance)
+		if err != nil {
+			c.syncMetricsCache.AddObject(
+				syncKey,
+				syncutil.Tags{
+					Kind:   instance.GetKind(),
+					Status: metrics.ErrorStatus,
+				},
+			)
+
+			return err
+		}
+
+		c.syncMetricsCache.AddObject(syncKey, syncutil.Tags{
+			Kind:   instance.GetKind(),
+			Status: metrics.ActiveStatus,
+		})
+		c.syncMetricsCache.AddKind(instance.GetKind())
+	}
+
+	c.tracker.ForData(instance.GroupVersionKind()).Observe(instance)
+
+	return nil
+}
+
+func (c *CacheManager) RemoveObject(ctx context.Context, instance *unstructured.Unstructured) error {
+	gvk := instance.GroupVersionKind()
+
+	if c.watchesGVK(gvk) {
+		if _, err := c.cfClient.RemoveData(ctx, instance); err != nil {
+			return err
+		}
+	}
+
+	// only delete from metrics map if the data removal was successful
+	c.syncMetricsCache.DeleteObject(syncutil.GetKeyForSyncMetrics(instance.GetNamespace(), instance.GetName()))
+	c.tracker.ForData(instance.GroupVersionKind()).CancelExpect(instance)
+
+	return nil
+}
+
+func (c *CacheManager) wipeData(ctx context.Context) error {
+	if _, err := c.cfClient.RemoveData(ctx, target.WipeData()); err != nil {
+		return err
+	}
+
+	// reset sync cache before sending the metric
+	c.syncMetricsCache.ResetCache()
+	c.syncMetricsCache.ReportSync()
+
+	return nil
+}
+
+func (c *CacheManager) ReportSyncMetrics() {
+	c.syncMetricsCache.ReportSync()
+}
+
+func (c *CacheManager) syncGVK(ctx context.Context, gvk schema.GroupVersionKind) error {
+	u := &unstructured.UnstructuredList{}
+	u.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   gvk.Group,
+		Version: gvk.Version,
+		Kind:    gvk.Kind + "List",
+	})
+
+	var err error
+	func() {
+		c.mu.Lock()
+		defer c.mu.Unlock()
+
+		// only call List if we are still watching the gvk.
+		if c.watchedSet.Contains(gvk) {
+			err = c.reader.List(ctx, u)
+		}
+	}()
+
+	if err != nil {
+		return fmt.Errorf("replaying data for %+v: %w", gvk, err)
+	}
+
+	for i := range u.Items {
+		if err := c.AddObject(ctx, &u.Items[i]); err != nil {
+			return fmt.Errorf("adding data for %+v: %w", gvk, err)
+		}
+	}
+
+	return nil
+}
+
+func (c *CacheManager) manageCache(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-c.backgroundManagementTicker.C:
+			func() {
+				c.mu.Lock()
+				defer c.mu.Unlock()
+
+				c.wipeCacheIfNeeded(ctx)
+
+				// spin up new goroutines to relist gvks as there has been a wipe
+				if c.needToList {
+					// stop any goroutines that were relisting before
+					// as we may no longer be interested in those gvks
+					close(c.relistStopChan)
+
+					// assume all gvks need to be relisted
+					gvksToRelist := c.gvksToSync.GVKs()
+
+					// clean state
+					c.needToList = false
+					c.relistStopChan = make(chan struct{})
```

Comment:
I think it's technically possible that we overwrite the `relistStopChan` with a new, empty channel before the `select` in `replayGVKs` has read the "stop" off of the channel.  As we've overwritten the shared variable, our stop event (which in its current state is just the closing of the channel) is never received.

Sequence of events:

1. Loop 1 of `manageCache` happens with many GVKs to list, potentially taking a long time.

1. Changes happen and `needToList == true`.

1. Loop 2 of `manageCache` kicks off.

1. Some GVK (`A`) is listing in Loop 1's `replayGVKs`

1. We `close(c.relistStopChan)`, meaning that Loop 1's `replayGVKs` should stop on the next iteration of its `select` statement.  This hasn't happened yet.

1. We overwrite the channel: `c.relistStopChan = make(chan struct{})`.

1. Loop 1's `replayGVKs`'s `select` statement finally checks the channel, but finds it open and empty (as it was recreated.

1. Loop 1's `replayGVKs` loop continues going through its stale list of GVKs.

Reply:
plz fact check me here @maxsmythe

Reply:
I think I buy that because we use a struct field here `c. relistStopChan` we can end up in a situation where we don't close the previous goroutine properly. Previous iterations here passed the stop ch as an argument to `replayGVKs` which could've prevented this situation iiuc.

Reply:
so we can revert back to a model where we do something like:

```
close(c.lastStopChRef)
stopCh := make(chan struct{})
c.lastStopChRef = stopCh
doStuff(stopCh)
```

Reply:
After thinking about this more, I think my advice has led this part of the code astray.

We need to have individual control over each instantiation of `go replayGVKs()`.  We _hope_ that there are never two instantiations running at once, but we don't actually have anything that prevents that from being so (as in, we are communicating over a channel and waiting on the other side to obey us).  Thus, we should assume that two goroutines may be spun up at the same time as an edge case.

In that scenario, we need to be able to communicate with the goroutines individually.  We want to stop the first one (via closing its channel) while allowing the second one to run AND maintaining a reference to the second one via a _new_ shared channel.

If we don't have per-goroutine channels, we lack the granularity required to accurately communicate.

So far, I've been resistant to closure-based management of data.  I have found it to be hard to grok and easy to break.  However, in this case I believe it's actually the best way to manage these channels.  It allows us to:

1. Maintain a reference to a channel across loops in `manageCache`.  This allows us to close the channel on our previous run of the loop.
1. Maintain a 1:1 relationship between channels and `replayGVKs` goroutines.  This allows us to close only the older goroutine in the case that we've spun up a new one before the first has been stopped.

Outside of managing this with closures, I think we'd have to make a more complicated structure that manages individual runs of `replayGVKs` and forces only one to run at a time.  That would be more heavyweight and I don't think is warranted considering we have a simpler solution through the use of a closure-variable.

Reply:
manageCache is only run as one goroutine (unless Start() is called multiple times, which we could guard against using sync.Once. Shouldn't happen, but adding it doesn't hurt).

b/c it's single threaded, and b/c of how the code in the for loop is ordered, stopChan will always be closed before it's overwritten.

Reply:
I think you are right, however, that using a struct-level variable to hold the stop chan can lead to a race condition:

* old replayGVKs is running
* stop chan is closed
* new replayGVKs is started
* stop chan is reassigned before old replayGVKs has had a chance to read it

General pattern for passing kill-signal-channels is:

* caller passes channel to child routine that it wants to kill as a parameter
* child routine reads from the passed parameter to grok whether it should stop
* caller maintains a reference to the routine in its scope

The sin we're committing here is that the child goroutine should not be reading stop chan from the struct. Whether we want to keep the stop channel as a struct variable or as a closure... I'd lean toward closure for simplicity, since it scopes the channel (and its definition) to the only place where its used.

Reply:
Thanks for the discussion y'all. The latest revision of the code reverts back to passing in the `stopCh` and moves away from using a struct field. I think we should be good now but feel free to reopen this conversation if I didn't quite hit the mark yet.